Deploying NASA Cryo cluster and hub

sgibson91 opened the pull request:
### Deployment-related changes in this PR

addresses https://github.com/2i2c-org/infrastructure/issues/1702

* Cluster files added:
  * Jsonnet files
  * SSH keys
  * Terraform files
  * cluster.yaml file
  * Deployer credentials
  * Grafana dashboards token
* Helm config files added:
  * Support components
  * Common helm config to be shared between hubs
  * Hub configs for staging and prod
* CI/CD files updated:
  * deploy grafana dashboards
  * deploy hubs
  * validate hubs

---

### Other changes in this PR

- Improvements to AWS deployment docs
- Improvement to AWS terraform code and related docs
  - Output the correct command to link the deployment credentials to the cluster
- Improvement to the AWS-related environment variables the deployer sets/unsets

sgibson91 commented:
I'm at the point in the documentation where I need to use terraform to generate a CI/CD key: https://infrastructure.2i2c.org/en/latest/howto/operate/new-cluster/aws.html#create-account-with-finely-scoped-permissions-for-automatic-deployment However, AWS terraform now seems to be doing _way_ more than just creating a CI/CD key and I don't know where that is documented.

The .tfvars file generate by `deployer generate-cluster` is not complete as was asked to provide `db_instance_identifier` on the command line. `db_enabled` should be set to `false` according to this code block: https://github.com/2i2c-org/infrastructure/blob/566c2a739feb116e7324402e9dceb50f8ca958db/terraform/aws/variables.tf#L66-L72

But my terraform plan is failing trying to set a root password for a mysql database - as far as I know, I don't want this feature enabled and it shouldn't be running!

```
$ tf plan -var-file=projects/nasa-cryo.tfvars -out=nasa-cryo-plan
data.aws_eks_cluster.cluster: Reading...
data.aws_partition.current: Reading...
data.aws_caller_identity.current: Reading...
data.aws_partition.current: Read complete after 0s [id=aws]
data.aws_caller_identity.current: Read complete after 0s [id=574251165169]
data.aws_eks_cluster.cluster: Read complete after 1s [id=nasa-cryo]
data.aws_security_group.cluster_nodes_shared_security_group: Reading...
data.aws_subnet.cluster_node_subnet: Reading...
data.aws_iam_policy_document.irsa_role_assume["staging"]: Reading...
data.aws_iam_policy_document.irsa_role_assume["prod"]: Reading...
data.aws_iam_policy_document.irsa_role_assume["prod"]: Read complete after 0s [id=1280574928]
data.aws_iam_policy_document.irsa_role_assume["staging"]: Read complete after 0s [id=3441862523]
data.aws_security_group.cluster_nodes_shared_security_group: Read complete after 0s [id=sg-094f411783a59613d]
data.aws_subnet.cluster_node_subnet: Read complete after 0s [id=subnet-0b5ab806454193c98]
╷
│ Error: Invalid index
│ 
│   on db.tf line 97, in provider "mysql":
│   97:   endpoint = aws_db_instance.db[0].endpoint
│     ├────────────────
│     │ aws_db_instance.db is empty tuple
│ 
│ The given key does not identify an element in this collection value: the collection has no elements.
╵
╷
│ Error: Invalid index
│ 
│   on db.tf line 99, in provider "mysql":
│   99:   username = aws_db_instance.db[0].username
│     ├────────────────
│     │ aws_db_instance.db is empty tuple
│ 
│ The given key does not identify an element in this collection value: the collection has no elements.
╵
╷
│ Error: Invalid index
│ 
│   on db.tf line 100, in provider "mysql":
│  100:   password = random_password.db_root_password[0].result
│     ├────────────────
│     │ random_password.db_root_password is empty tuple
│ 
│ The given key does not identify an element in this collection value: the collection has no elements.
╵
```

yuvipanda commented on terraform/aws/projects/nasa-cryo.tfvars:
```diff
@@ -0,0 +1,32 @@
+
+# Sarah added this value because terraform asked for it to be provided on the
+# command line when she executed `tf plan`
+db_instance_identifier = "nasa-cryo"
```
this can be removed now

sgibson91 commented:
Output of `terraform plan`

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create
 <= read (data resources)

Terraform will perform the following actions:

  # data.aws_iam_policy_document.bucket_access["prod.scratch"] will be read during apply
  # (config refers to values not yet known)
 <= data "aws_iam_policy_document" "bucket_access" {
      + id   = (known after apply)
      + json = (known after apply)

      + statement {
          + actions   = [
              + "s3:*",
            ]
          + effect    = "Allow"
          + resources = [
              + (known after apply),
              + (known after apply),
            ]

          + principals {
              + identifiers = [
                  + (known after apply),
                ]
              + type        = "AWS"
            }
        }
    }

  # data.aws_iam_policy_document.bucket_access["staging.scratch-staging"] will be read during apply
  # (config refers to values not yet known)
 <= data "aws_iam_policy_document" "bucket_access" {
      + id   = (known after apply)
      + json = (known after apply)

      + statement {
          + actions   = [
              + "s3:*",
            ]
          + effect    = "Allow"
          + resources = [
              + (known after apply),
              + (known after apply),
            ]

          + principals {
              + identifiers = [
                  + (known after apply),
                ]
              + type        = "AWS"
            }
        }
    }

  # aws_efs_file_system.homedirs will be created
  + resource "aws_efs_file_system" "homedirs" {
      + arn                     = (known after apply)
      + availability_zone_id    = (known after apply)
      + availability_zone_name  = (known after apply)
      + creation_token          = (known after apply)
      + dns_name                = (known after apply)
      + encrypted               = (known after apply)
      + id                      = (known after apply)
      + kms_key_id              = (known after apply)
      + number_of_mount_targets = (known after apply)
      + owner_id                = (known after apply)
      + performance_mode        = (known after apply)
      + size_in_bytes           = (known after apply)
      + tags                    = {
          + "Name" = "hub-homedirs"
        }
      + tags_all                = {
          + "Name" = "hub-homedirs"
        }
      + throughput_mode         = "bursting"
    }

  # aws_efs_mount_target.homedirs will be created
  + resource "aws_efs_mount_target" "homedirs" {
      + availability_zone_id   = (known after apply)
      + availability_zone_name = (known after apply)
      + dns_name               = (known after apply)
      + file_system_arn        = (known after apply)
      + file_system_id         = (known after apply)
      + id                     = (known after apply)
      + ip_address             = (known after apply)
      + mount_target_dns_name  = (known after apply)
      + network_interface_id   = (known after apply)
      + owner_id               = (known after apply)
      + security_groups        = [
          + "sg-094f411783a59613d",
        ]
      + subnet_id              = "subnet-0b5ab806454193c98"
    }

  # aws_iam_access_key.continuous_deployer will be created
  + resource "aws_iam_access_key" "continuous_deployer" {
      + create_date                    = (known after apply)
      + encrypted_secret               = (known after apply)
      + encrypted_ses_smtp_password_v4 = (known after apply)
      + id                             = (known after apply)
      + key_fingerprint                = (known after apply)
      + secret                         = (sensitive value)
      + ses_smtp_password_v4           = (sensitive value)
      + status                         = "Active"
      + user                           = "hub-continuous-deployer"
    }

  # aws_iam_role.irsa_role["prod"] will be created
  + resource "aws_iam_role" "irsa_role" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRoleWithWebIdentity"
                      + Condition = {
                          + StringEquals = {
                              + "oidc.eks.us-west-2.amazonaws.com/id/3D5422BE1F96E1111E1C0B882B64A16B:sub" = "system:serviceaccount:prod:user-sa"
                            }
                        }
                      + Effect    = "Allow"
                      + Principal = {
                          + Federated = "arn:aws:iam::574251165169:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/3D5422BE1F96E1111E1C0B882B64A16B"
                        }
                      + Sid       = ""
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = "nasa-cryo-prod"
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags_all              = (known after apply)
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # aws_iam_role.irsa_role["staging"] will be created
  + resource "aws_iam_role" "irsa_role" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRoleWithWebIdentity"
                      + Condition = {
                          + StringEquals = {
                              + "oidc.eks.us-west-2.amazonaws.com/id/3D5422BE1F96E1111E1C0B882B64A16B:sub" = "system:serviceaccount:staging:user-sa"
                            }
                        }
                      + Effect    = "Allow"
                      + Principal = {
                          + Federated = "arn:aws:iam::574251165169:oidc-provider/oidc.eks.us-west-2.amazonaws.com/id/3D5422BE1F96E1111E1C0B882B64A16B"
                        }
                      + Sid       = ""
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = "nasa-cryo-staging"
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags_all              = (known after apply)
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # aws_iam_user.continuous_deployer will be created
  + resource "aws_iam_user" "continuous_deployer" {
      + arn           = (known after apply)
      + force_destroy = false
      + id            = (known after apply)
      + name          = "hub-continuous-deployer"
      + path          = "/"
      + tags_all      = (known after apply)
      + unique_id     = (known after apply)
    }

  # aws_iam_user_policy.continuous_deployer will be created
  + resource "aws_iam_user_policy" "continuous_deployer" {
      + id     = (known after apply)
      + name   = "eks-readonly"
      + policy = jsonencode(
            {
              + Statement = [
                  + {
                      + Action   = "eks:DescribeCluster"
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + user   = "hub-continuous-deployer"
    }

  # aws_s3_bucket.user_buckets["scratch"] will be created
  + resource "aws_s3_bucket" "user_buckets" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "nasa-cryo-scratch"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + cors_rule {
          + allowed_headers = (known after apply)
          + allowed_methods = (known after apply)
          + allowed_origins = (known after apply)
          + expose_headers  = (known after apply)
          + max_age_seconds = (known after apply)
        }

      + grant {
          + id          = (known after apply)
          + permissions = (known after apply)
          + type        = (known after apply)
          + uri         = (known after apply)
        }

      + lifecycle_rule {
          + abort_incomplete_multipart_upload_days = (known after apply)
          + enabled                                = (known after apply)
          + id                                     = (known after apply)
          + prefix                                 = (known after apply)
          + tags                                   = (known after apply)

          + expiration {
              + date                         = (known after apply)
              + days                         = (known after apply)
              + expired_object_delete_marker = (known after apply)
            }

          + noncurrent_version_expiration {
              + days = (known after apply)
            }

          + noncurrent_version_transition {
              + days          = (known after apply)
              + storage_class = (known after apply)
            }

          + transition {
              + date          = (known after apply)
              + days          = (known after apply)
              + storage_class = (known after apply)
            }
        }

      + logging {
          + target_bucket = (known after apply)
          + target_prefix = (known after apply)
        }

      + object_lock_configuration {
          + object_lock_enabled = (known after apply)

          + rule {
              + default_retention {
                  + days  = (known after apply)
                  + mode  = (known after apply)
                  + years = (known after apply)
                }
            }
        }

      + replication_configuration {
          + role = (known after apply)

          + rules {
              + delete_marker_replication_status = (known after apply)
              + id                               = (known after apply)
              + prefix                           = (known after apply)
              + priority                         = (known after apply)
              + status                           = (known after apply)

              + destination {
                  + account_id         = (known after apply)
                  + bucket             = (known after apply)
                  + replica_kms_key_id = (known after apply)
                  + storage_class      = (known after apply)

                  + access_control_translation {
                      + owner = (known after apply)
                    }

                  + metrics {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }

                  + replication_time {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }
                }

              + filter {
                  + prefix = (known after apply)
                  + tags   = (known after apply)
                }

              + source_selection_criteria {
                  + sse_kms_encrypted_objects {
                      + enabled = (known after apply)
                    }
                }
            }
        }

      + server_side_encryption_configuration {
          + rule {
              + bucket_key_enabled = (known after apply)

              + apply_server_side_encryption_by_default {
                  + kms_master_key_id = (known after apply)
                  + sse_algorithm     = (known after apply)
                }
            }
        }

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }

      + website {
          + error_document           = (known after apply)
          + index_document           = (known after apply)
          + redirect_all_requests_to = (known after apply)
          + routing_rules            = (known after apply)
        }
    }

  # aws_s3_bucket.user_buckets["scratch-staging"] will be created
  + resource "aws_s3_bucket" "user_buckets" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "nasa-cryo-scratch-staging"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + cors_rule {
          + allowed_headers = (known after apply)
          + allowed_methods = (known after apply)
          + allowed_origins = (known after apply)
          + expose_headers  = (known after apply)
          + max_age_seconds = (known after apply)
        }

      + grant {
          + id          = (known after apply)
          + permissions = (known after apply)
          + type        = (known after apply)
          + uri         = (known after apply)
        }

      + lifecycle_rule {
          + abort_incomplete_multipart_upload_days = (known after apply)
          + enabled                                = (known after apply)
          + id                                     = (known after apply)
          + prefix                                 = (known after apply)
          + tags                                   = (known after apply)

          + expiration {
              + date                         = (known after apply)
              + days                         = (known after apply)
              + expired_object_delete_marker = (known after apply)
            }

          + noncurrent_version_expiration {
              + days = (known after apply)
            }

          + noncurrent_version_transition {
              + days          = (known after apply)
              + storage_class = (known after apply)
            }

          + transition {
              + date          = (known after apply)
              + days          = (known after apply)
              + storage_class = (known after apply)
            }
        }

      + logging {
          + target_bucket = (known after apply)
          + target_prefix = (known after apply)
        }

      + object_lock_configuration {
          + object_lock_enabled = (known after apply)

          + rule {
              + default_retention {
                  + days  = (known after apply)
                  + mode  = (known after apply)
                  + years = (known after apply)
                }
            }
        }

      + replication_configuration {
          + role = (known after apply)

          + rules {
              + delete_marker_replication_status = (known after apply)
              + id                               = (known after apply)
              + prefix                           = (known after apply)
              + priority                         = (known after apply)
              + status                           = (known after apply)

              + destination {
                  + account_id         = (known after apply)
                  + bucket             = (known after apply)
                  + replica_kms_key_id = (known after apply)
                  + storage_class      = (known after apply)

                  + access_control_translation {
                      + owner = (known after apply)
                    }

                  + metrics {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }

                  + replication_time {
                      + minutes = (known after apply)
                      + status  = (known after apply)
                    }
                }

              + filter {
                  + prefix = (known after apply)
                  + tags   = (known after apply)
                }

              + source_selection_criteria {
                  + sse_kms_encrypted_objects {
                      + enabled = (known after apply)
                    }
                }
            }
        }

      + server_side_encryption_configuration {
          + rule {
              + bucket_key_enabled = (known after apply)

              + apply_server_side_encryption_by_default {
                  + kms_master_key_id = (known after apply)
                  + sse_algorithm     = (known after apply)
                }
            }
        }

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }

      + website {
          + error_document           = (known after apply)
          + index_document           = (known after apply)
          + redirect_all_requests_to = (known after apply)
          + routing_rules            = (known after apply)
        }
    }

  # aws_s3_bucket_lifecycle_configuration.user_bucket_expiry["scratch"] will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "user_bucket_expiry" {
      + bucket = "nasa-cryo-scratch"
      + id     = (known after apply)

      + rule {
          + id     = "delete-after-expiry"
          + status = "Enabled"

          + expiration {
              + days                         = 7
              + expired_object_delete_marker = (known after apply)
            }
        }
    }

  # aws_s3_bucket_lifecycle_configuration.user_bucket_expiry["scratch-staging"] will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "user_bucket_expiry" {
      + bucket = "nasa-cryo-scratch-staging"
      + id     = (known after apply)

      + rule {
          + id     = "delete-after-expiry"
          + status = "Enabled"

          + expiration {
              + days                         = 7
              + expired_object_delete_marker = (known after apply)
            }
        }
    }

  # aws_s3_bucket_policy.user_bucket_access["prod.scratch"] will be created
  + resource "aws_s3_bucket_policy" "user_bucket_access" {
      + bucket = (known after apply)
      + id     = (known after apply)
      + policy = (known after apply)
    }

  # aws_s3_bucket_policy.user_bucket_access["staging.scratch-staging"] will be created
  + resource "aws_s3_bucket_policy" "user_bucket_access" {
      + bucket = (known after apply)
      + id     = (known after apply)
      + policy = (known after apply)
    }

Plan: 13 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + continuous_deployer_creds = (sensitive value)
  + db_helm_config            = (sensitive value)
  + kubernetes_sa_annotations = {
      + prod    = (known after apply)
      + staging = (known after apply)
    }
  + nfs_server_dns            = (known after apply)
```

sgibson91 commented:
@yuvipanda The next bug that the deployer credentials created by terraform don't seem to work

```
$ python deployer use-cluster-credentials nasa-cryo

An error occurred (UnrecognizedClientException) when calling the DescribeCluster operation: The security token included in the request is invalid
Traceback (most recent call last):
  File "/usr/local/Caskroom/miniconda/base/envs/infrastructure/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Caskroom/miniconda/base/envs/infrastructure/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/sgibson/source/github/2i2c-org/infrastructure/deployer/__main__.py", line 7, in <module>
    cli.main()
  File "/Users/sgibson/source/github/2i2c-org/infrastructure/deployer/cli.py", line 199, in main
    use_cluster_credentials(args.cluster_name)
  File "/Users/sgibson/source/github/2i2c-org/infrastructure/deployer/deploy_actions.py", line 58, in use_cluster_credentials
    with cluster.auth():
  File "/usr/local/Caskroom/miniconda/base/envs/infrastructure/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/Users/sgibson/source/github/2i2c-org/infrastructure/deployer/cluster.py", line 29, in auth
    yield from self.auth_aws()
  File "/Users/sgibson/source/github/2i2c-org/infrastructure/deployer/cluster.py", line 180, in auth_aws
    subprocess.check_call(
  File "/usr/local/Caskroom/miniconda/base/envs/infrastructure/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['aws', 'eks', 'update-kubeconfig', '--name=nasa-cryo', '--region=us-west-2']' returned non-zero exit status 254.
```

I'm trying to deploy the support chart but can't. 

I can successfully authenticate against the cluster myself using my environment variables generated from following [these docs](https://infrastructure.2i2c.org/en/latest/topic/cloud-auth.html#access-aws-from-your-terminal)

```
$ aws eks update-kubeconfig --name=nasa-cryo --region=us-west-2
Updated context arn:aws:eks:us-west-2:574251165169:cluster/nasa-cryo in /Users/sgibson/.kube/config
```

But whatever gets exported from `terraform output -raw continuous_deployer_creds > ../../config/clusters/nasa-cryo/deployer-credentials.secret.json` is not working

GeorgianaElena commented:
@sgibson91, can you check if the rows in the json that stores the creds are tabs instead of spaces?
I believe I once needed to manually change them

sgibson91 commented:
> @sgibson91, can you check if the rows in the json that stores the creds are tabs instead of spaces? I believe I once needed to manually change them.
> 
> Update: Ah, and I believe the sops enc output needs this indentation. I doesn't matter if you have the credentials indented with spaces originally, before encryption. I think `sops` is the one that turns them into tabs.

This is all correct. But also, the deployer opens json files with the json library (instead of yaml) precisely because of the hard tabs. I will dig out the PR where Yuvi reintroduced this.

damianavila commented:
> But whatever gets exported from terraform output -raw continuous_deployer_creds > ../../config/clusters/nasa-cryo/deployer-credentials.secret.json is not working

AFAIR, the terraform pieces create the deployer IAM user but you still need to give it access manually: https://infrastructure.2i2c.org/en/latest/howto/operate/new-cluster/aws.html#grant-access-to-other-users. Maybe this is the underlying issue?

yuvipanda commented:
+1 to what @damianavila linked to. Might be the cause. Unfortunately only you can do this, @sgibson91 because of https://infrastructure.2i2c.org/en/latest/howto/operate/new-cluster/aws.html?highlight=access#grant-access-to-other-users.

sgibson91 commented:
I did that though?

yuvipanda commented:
@sgibson91 oh damn, that sucks. Did you also do https://infrastructure.2i2c.org/en/latest/howto/operate/new-cluster/aws.html?highlight=access#grant-access-to-other-users? I can't seem to get access.

sgibson91 commented:
<img width="1408" alt="Screenshot 2022-10-14 at 19 31 05" src="https://user-images.githubusercontent.com/44771837/195916555-61b1b270-ca30-4a43-88df-46ea0038d592.png">

> Did you also do [infrastructure.2i2c.org/en/latest/howto/operate/new-cluster/aws.html?highlight=access#grant-access-to-other-users](https://infrastructure.2i2c.org/en/latest/howto/operate/new-cluster/aws.html?highlight=access#grant-access-to-other-users)? I can't seem to get access.

Yes, I looped through everyone's user names that I found here: https://us-east-1.console.aws.amazon.com/iamv2/home?region=us-east-1#/groups/details/tech?section=users

yuvipanda commented:
@sgibson91 ok that's super strange! I'll investigate. THANK YOU!

sgibson91 commented:
<img width="1289" alt="Screenshot 2022-10-14 at 19 33 19" src="https://user-images.githubusercontent.com/44771837/195916916-150259f4-ec5a-4a39-b2fc-fa9cac44f5cc.png">

yuvipanda commented:
@sgibson91 ah, my credentials had just expired i think. I can now reproduce your error, getting:

`An error occurred (UnrecognizedClientException) when calling the DescribeCluster operation: The security token included in the request is invalid`

I'll look at that now.

yuvipanda commented:
@sgibson91 which terraform workspace is this in? I don't see a nasa-cryo one in:

```
terraform workspace list
  default
  2i2c-uk
  allen-swdb
  awi-ciroh
  callysto
* carbonplan-aws
  cloudbank
  justiceinnovationlab
  leap
  linked-earth
  m2lines
  meom-ige
  openscapes
  pilot-hubs
  uhackweeks
  utoronto
```

yuvipanda commented:
@sgibson91 no worries, that happens to me all the time! I'll look in default. We can delete and recreate in a new workspace before going live. I am grateful you are working on the docs as you're going too so we can streamline this <3

yuvipanda commented:
If I take the current key and secret and set them as env vars, and run `aws eks describe-cluster --name nasa-cryo --region=us-west-2`, I do get the correct output. So curious why our deployer doesn't.

yuvipanda commented:
I ran ` k -n kube-system get cm aws-auth -o yaml` and found two entries for `hub-continuous-deployer` and deleted one. That had no effect.

yuvipanda commented:
`aws eks update-kubeconfig --name nasa-cryo --region=us-west-2` works as me, but not as the deployer. I use `aws sts get-caller-identity` to validate whose credentials are being used to make calls

yuvipanda commented:
@sgibson91 aha, found it! So turns out that our engineer's users are in the org id `746653422107` which is our SSO org, but the hub-deployer user is actually in the same org as the cluster itself, which in this case was `574251165169`. Once I switched that up, it worked fine! I don't think we have this clearly documented anywhere, so I've added a `terraform output` that will provide the exact correct command to run and added that to our docs, and pushed the commit to this PR

sgibson91 commented:
Thank you @yuvipanda! 🚀

sgibson91 commented on docs/howto/operate/new-cluster/aws.md:
```diff
@@ -222,6 +232,7 @@ Get the address a hub on this cluster should use for connecting to NFS with
 `terraform output nfs_server_dns`, and set it in the hub's config under
 `nfs.pv.serverIP` (nested under `basehub` when necessary) in the appropriate
 `<hub>.values.yaml` file.
+>>>>>>> upstream/master
```
```suggestion
```

sgibson91 commented:
@yuvipanda I'm still getting an error trying to run `python deployer use-cluster-credentials` 😭 😭 😭 

While I was at it, I terraform-destroyed everything I had deployed previously, created a new terraform workspace, pulled your changes, then redeployed all the terraform stuff into the new workspace. Even using the tarraform command output you provided, and updating the exported deployer credentials, I'm still stuck!

sgibson91 commented:
Do I have to redeploy the whole eksctl cluster?

yuvipanda commented:
@sgibson91 so I opened a new shell, and tried:

```
➜ python3 deployer use-cluster-credentials nasa-cryo                          
Added new context arn:aws:eks:us-west-2:574251165169:cluster/nasa-cryo to /var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/tmpmovvgznk

🐍 infrastructure g:nasa-cryo gcp:ucb-datahub-2018 k:arn:aws:eks:us-west-2:574251165169:cluster/nasa-cryo/ arm64 
➜ k get node
NAME                                          STATUS   ROLES    AGE     VERSION
ip-192-168-21-95.us-west-2.compute.internal   Ready    <none>   4d17h   v1.22.12-eks-ba74326
```

so it does work for me.

However, in another terminal, with some older `AWS` tokens set, I see:

```
➜ python3 deployer use-cluster-credentials nasa-cryo

An error occurred (UnrecognizedClientException) when calling the DescribeCluster operation: The security token included in the request is invalid
```

Can you check what env vars are set prior?

yuvipanda commented:
@sgibson91 I made the deployer slightly more expansive in the variables it unsets / restores wrt AWS! So it *should* just work without other terminals now. I pushed the commit to this PR.

yuvipanda commented:
@sgibson91 I don't think you need to setup a new eksctl cluster.

sgibson91 commented:
Thank you @yuvipanda - this finally works 🚀 I had shut down my laptop over the weekend, so yesterday it should have been a completely new shell will new env vars generated from the awsapps login page. So I don't really know what was causing it to b0rk other than that me and the VSCode integrated terminal are becoming less and less friendly.

sgibson91 commented:
If I could get a sanity check that I have enabled all the features requested in https://github.com/2i2c-org/infrastructure/issues/1702 in the `common.values.yaml` file before I setup authentication and start deploying hubs tomorrow, I would be very grateful

- https://github.com/2i2c-org/infrastructure/pull/1768/files#diff-67a27eea3a5155ba5cdd45181ac7ec6f2870b6f1426ace75b86e60359b082880

### Sidebar

I had a conversation with @damianavila today that it would be useful now to have some `basehub-common-template.values.yaml` and `daskhub-common-template.values.yaml` files that have the bare minimum required to copy from, since it is becoming more difficult to determine what is and what isn't customised just by inspecting _other_ hub's values files.

I don't even need a fancy `deployer generate-hub` command - some files in the repo would suffice :)

yuvipanda commented on config/clusters/nasa-cryo/common.values.yaml:
```diff
@@ -0,0 +1,122 @@
+          - name: home
+            mountPath: /home/jovyan/shared
+            subPath: _shared
+            readOnly: false
```
```suggestion
            readOnly: true
```

I think by default, we should make shared be readonly (and shared-readwrite available to admins as read-write). I think for some specific hub we made shared read-write for everyone, but don't think that should be default

yuvipanda commented on docs/howto/operate/new-cluster/aws.md:
```diff
@@ -184,9 +161,12 @@ have least amount of permissions possible.
    this can be a serious security leak!
 
-2. Grant the freshly created IAM user access to the kubernetes cluster.
+2. Grant the freshly created IAM user access to the kubernetes cluster. As this requires
```
We should renumber these. This should be 5? I guess it was already wrong :D

sgibson91 replied:
All the numbering looks fine to me on the preview? https://2i2c-pilot-hubs--1768.org.readthedocs.build/en/1768/howto/operate/new-cluster/aws.html

yuvipanda reviewed (approved):
Two minor comments otherwise LGMT @sgibson91

colliand commented:
I am 🤩 that this hub will soon launch.

sgibson91 commented:
Two hubs are new up and running so I am marking this PR ready for review

* https://staging.cryointhecloud.2i2c.cloud
* https://cryointhecloud.2i2c.cloud

sgibson91 commented:
For some reason, `deployer run-hub-health-check` fails for these hubs. The deployment service check can't create a user server

```
------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------
Starting hub https://staging.cryointhecloud.2i2c.cloud health validation...
Running dask_test_notebook.ipynb test notebook...
Hub https://staging.cryointhecloud.2i2c.cloud not healthy! Stopping further deployments. Exception was jupyterhub server creation timeout=360 [s].
--------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------
ERROR    jhub_client.api:api.py:119 jupyterhub server creation timeout=360 [s]
============================================================================== short test summary info ==============================================================================
FAILED deployer/tests/test_hub_health.py::test_hub_healthy - TimeoutError: jupyterhub server creation timeout=360 [s]
1 failed in 364.77s (0:06:04)
Health check failed!
```

sgibson91 commented:
Now my spawn is just hanging

<img width="813" alt="Screenshot 2022-10-19 at 15 35 17" src="https://user-images.githubusercontent.com/44771837/196721667-1de27848-bad3-4919-90e4-bbb298e5c5e6.png">

```
$ k describe pod jupyter-sgibson91
[...]
Events:
  Type     Reason            Age                  From               Message
  ----     ------            ----                 ----               -------
  Warning  FailedScheduling  14s (x8 over 7m25s)  default-scheduler  0/1 nodes are available: 1 node(s) didn't match Pod's node affinity/selector.
```

I think it might be trying to auto-scale but can't? Not seeing the kind of scaling failed messages I'd expect though.

```
[I 2022-10-19 14:28:17.997 JupyterHub app:3095] Private Hub API connect url http://hub:8081/hub/
[I 2022-10-19 14:28:17.997 JupyterHub app:3104] Starting managed service jupyterhub-idle-culler
[I 2022-10-19 14:28:17.997 JupyterHub service:385] Starting service 'jupyterhub-idle-culler': ['python3', '-m', 'jupyterhub_idle_culler', '--url=http://localhost:8081/hub/api', '--timeout=3600', '--cull-every=600', '--concurrency=10']
[I 2022-10-19 14:28:17.999 JupyterHub service:133] Spawning python3 -m jupyterhub_idle_culler --url=http://localhost:8081/hub/api --timeout=3600 --cull-every=600 --concurrency=10
[I 2022-10-19 14:28:18.004 JupyterHub app:3104] Starting managed service configurator at http://configurator:10101
[I 2022-10-19 14:28:18.004 JupyterHub service:385] Starting service 'configurator': ['python3', '-m', 'jupyterhub_configurator.app', '--Configurator.config_file=/usr/local/etc/jupyterhub-configurator/jupyterhub_configurator_config.py']
[I 2022-10-19 14:28:18.007 JupyterHub service:133] Spawning python3 -m jupyterhub_configurator.app --Configurator.config_file=/usr/local/etc/jupyterhub-configurator/jupyterhub_configurator_config.py
[I 2022-10-19 14:28:18.132 JupyterHub log:186] 200 GET /hub/api/ (jupyterhub-idle-culler@127.0.0.1) 17.04ms
[I 2022-10-19 14:28:18.145 JupyterHub log:186] 200 GET /hub/api/users?state=[secret] (jupyterhub-idle-culler@127.0.0.1) 11.70ms
[I 2022-10-19 14:28:19.025 JupyterHub app:3113] Adding external service dask-gateway at http://traefik-staging-dask-gateway.staging
[I 2022-10-19 14:28:19.028 JupyterHub app:3113] Adding external service hub-health
[I 2022-10-19 14:28:19.031 JupyterHub app:3162] JupyterHub is now running, internal Hub API at http://hub:8081/hub/
[I 2022-10-19 14:28:32.912 JupyterHub log:186] 200 GET /hub/metrics (@192.168.12.26) 4.12ms
User sgibson91 is part of teams WhitakerLab:labmembers 2i2c-org:tech-team WhitakerLab:turingregcollabs
Allowing profile Small: m5.large for user sgibson91 based on team membership
Allowing profile Medium: m5.xlarge for user sgibson91 based on team membership
Allowing profile Large: m5.2xlarge for user sgibson91 based on team membership
Allowing profile Huge: m5.8xlarge for user sgibson91 based on team membership
[I 2022-10-19 14:29:13.554 JupyterHub log:186] 200 GET /hub/spawn/sgibson91 (sgibson91@192.168.21.95) 58.97ms
[I 2022-10-19 14:29:13.555 JupyterHub reflector:274] watching for pods with label selector='component=singleuser-server' in namespace staging
[I 2022-10-19 14:29:13.560 JupyterHub reflector:274] watching for events with field selector='involvedObject.kind=Pod' in namespace staging
[I 2022-10-19 14:29:15.693 JupyterHub provider:651] Creating oauth client jupyterhub-user-sgibson91
[I 2022-10-19 14:29:15.709 JupyterHub log:186] 302 POST /hub/spawn/sgibson91 -> /hub/spawn-pending/sgibson91 (sgibson91@192.168.21.95) 36.51ms
[I 2022-10-19 14:29:15.727 JupyterHub spawner:2469] Attempting to create pod jupyter-sgibson91, with timeout 3
[I 2022-10-19 14:29:15.884 JupyterHub pages:394] sgibson91 is pending spawn
[I 2022-10-19 14:29:15.888 JupyterHub log:186] 200 GET /hub/spawn-pending/sgibson91 (sgibson91@192.168.21.95) 6.81ms
[I 2022-10-19 14:29:32.912 JupyterHub log:186] 200 GET /hub/metrics (@192.168.12.26) 4.34ms
[I 2022-10-19 14:30:32.912 JupyterHub log:186] 200 GET /hub/metrics (@192.168.12.26) 4.33ms
[I 2022-10-19 14:31:16.105 JupyterHub log:186] 302 GET / -> /hub/ (@192.168.21.95) 0.68ms
[I 2022-10-19 14:31:16.171 JupyterHub log:186] 302 GET /hub/ -> /hub/login?next=%2Fhub%2F (@192.168.21.95) 0.92ms
[I 2022-10-19 14:31:16.252 JupyterHub log:186] 200 GET /hub/login?next=/hub/ (@192.168.21.95) 15.87ms
[I 2022-10-19 14:31:32.913 JupyterHub log:186] 200 GET /hub/metrics (@192.168.12.26) 5.09ms
[I 2022-10-19 14:32:32.912 JupyterHub log:186] 200 GET /hub/metrics (@192.168.12.26) 4.92ms
[I 2022-10-19 14:33:32.912 JupyterHub log:186] 200 GET /hub/metrics (@192.168.12.26) 5.03ms
```

sgibson91 commented:
Exactly which quotas are we supposed to request increases for?

![Screenshot 2022-10-19 at 15 40 43](https://user-images.githubusercontent.com/44771837/196723036-aed9aa4d-3f24-490a-af79-471a34e23407.png)

GeorgianaElena commented:
Pinging @damianavila in case he might be able to help with ☝🏼

GeorgianaElena commented on eksctl/nasa-cryo.jsonnet:
```diff
@@ -0,0 +1,109 @@
+    },
+    availabilityZones: masterAzs,
+    iam: {
+        withOIDC: true,
```
@sgibson91, I might be far from truth, but I noticed https://github.com/2i2c-org/infrastructure/blob/bf6ae45b97d27cf303c68379c35adf9d493097cf/eksctl/uwhackweeks.jsonnet#L62-L66 that is missing here.

Shouldn't we add the `nasa-cryo-staging` and the prod one here and then list the correspoding service account in singleuser, similar to https://github.com/2i2c-org/infrastructure/pull/1768/commits/559c1e384635c75d55c791380d40fb93a0160dc1 ?

GeorgianaElena replied:
I can still see the `cloud-user-sa` in the logs `serviceaccount \"cloud-user-sa\"` for some reason
````

sgibson91 replied:
I removed it from the hub config, not here. Am I supposed to?

sgibson91 replied:
I've never used the jsonnet files before so I am clueless and completely dependent on @yuvipanda's `generate-cluster` command to "do the right thing" here

sgibson91 replied:
>  logs serviceaccount \"cloud-user-sa\" for some reason

Did you check the prod logs? I probably only ran the update against staging to debug

sgibson91 replied:
I updated both hubs just now

GeorgianaElena replied:
Yes, I was in the prod hub sorry for the cofusion :(

sgibson91 replied:
I added the prod/staging annotations in their respective `<hub>.values.yaml` files, like:

https://github.com/2i2c-org/infrastructure/pull/1768/files#diff-5204ecf6a357bcb3d7cff3005b283cf99acd6af7b4ff054e6ab13df64a198e56R2-R4

I have no idea if they are the same/different to `cloud-user-sa`

GeorgianaElena commented on eksctl/nasa-cryo.jsonnet:
```diff
@@ -0,0 +1,109 @@
+            // instanceTypes always have a .
+            name: "dask-%s" % std.strReplace(n.instancesDistribution.instanceTypes[0], ".", "-"),
+            availabilityZones: [nodeAz],
+            minSize: 0,
```
Wondering what would happen if we would force the creation of one notebook node here. Would the server be allocated to it if it would exist? Or is it rather a node creation issue

sgibson91 replied:
That worked!

yuvipanda commented:
@sgibson91 @GeorgianaElena on EKS, the cluster autoscaler needs to be explicitly enabled by us in our support chart. I did so in https://github.com/2i2c-org/infrastructure/pull/1768/commits/fedcfb2c2d7a10bfa06d95fb7adfaa40bea59611 and it's on now!

The `cloud-user-sa` is also openscapes and carbonplan specific, and predates the work in https://infrastructure.2i2c.org/en/latest/howto/features/cloud-access.html that involves annotations and stuff. 

We should *definitely* have templates here for sure 100%, including for support charts.

yuvipanda commented:
I opened https://github.com/2i2c-org/infrastructure/issues/1800 to get rid of cloud-user-sa (

yuvipanda commented:
As for quotas, we need mostly EC2 quotas (https://us-east-1.console.aws.amazon.com/servicequotas/home/services/ec2/quotas) for new nodes to come up. In particular:

`All Standard (A, C, D, H, I, M, R, T, Z) Spot Instance Requests` is what is used for dask instances (as they are spot instances), and `Running On-Demand Standard (A, C, D, H, I, M, R, T, Z) instances` is what is used for notebook and core instances. The values are 'total CPUs', so bigger nodes consume more quota.

sgibson91 commented:
Thanks @yuvipanda - I updated our docs around cluster-autoscaler and quotas as well. It's probably fine for now, and I'll give flow a rethink when I begin tackling https://github.com/2i2c-org/infrastructure/issues/1757

yuvipanda reviewed (approved):
Yay awesome!

sgibson91 commented:
Will hold off merging this until Tasha has set the CNAME they would like to use :)

sgibson91 commented:
I updated the teams to match the capitalisation in the _slugs_ not the _display names_: https://github.com/2i2c-org/infrastructure/issues/1702#issuecomment-1285863393

I also updated the domains so the hubs are available at the desired CNAMEs of the community.

Merging now!
